### PR TITLE
[RNMobile] Upgrade to RN 0.64

### PR DESCRIPTION
**Pull request by dratwas**
## Description
Upgrade React Native to version 0.64.0 (See [RN Upgrade Helper](https://react-native-community.github.io/upgrade-helper/?from=0.61.5&to=0.64.0))

Related PRs:

- Gutenberg-Mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3430
- WPiOS: https://github.com/wordpress-mobile/WordPress-iOS/pull/16562
- WPAndroid: https://github.com/wordpress-mobile/WordPress-Android/pull/14702
- react-native-svg PR: https://github.com/wordpress-mobile/react-native-svg/pull/12
- react-native-keyboard-aware-scroll-view PR: https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/pull/12
- react-native-gesture-handler PR: https://github.com/wordpress-mobile/react-native-gesture-handler/pull/1
- react-native-video: https://github.com/wordpress-mobile/react-native-video/pull/10
- react-native-linear-gradient: https://github.com/wordpress-mobile/react-native-linear-gradient/pull/1
- react-native-hsv-color-picker: https://github.com/wordpress-mobile/react-native-hsv-color-picker/pull/1
- react-native-safe-area-context upgrade PR: https://github.com/wordpress-mobile/react-native-safe-area-context/pull/2
- react-native-slider: https://github.com/wordpress-mobile/react-native-slider/pull/6
- react-native-get-random-values: https://github.com/wordpress-mobile/react-native-get-random-values/pull/1
- react-native-masked-view: https://github.com/wordpress-mobile/react-native-masked-view/pull/2
- react-native-screens: https://github.com/wordpress-mobile/react-native-screens/pull/2
- react-native-reanimated: https://github.com/wordpress-mobile/react-native-reanimated/pull/2
- react-native-prompt-android: https://github.com/wordpress-mobile/react-native-prompt-android/pull/3

## How has this been tested?

To test: Gutenberg Editor should work as expected.
Try to run all the manual tests.

Remaining Tasks:
- [x] `rnmobile/try/upgrade-0-64` branch is updated to 0.64.0-rc3 and needs to be updated to 0.64.0 stable. 
- [x] `rnmobile/try/upgrade-0-64` needs to be updated with latest gutenberg trunk which contains some larger updates for handling WPAndroid composite builds
- [x] iOS Demo App needs to be fixed on `rnmobile/try/upgrade-0-64`
- [x] Need to create relevant PRs in other Repos following model of https://github.com/wordpress-mobile/gutenberg-mobile/pull/2580
- [x] Re-target to [Gutenberg Web React 17 update branch](https://href.li/?https://github.com/WordPress/gutenberg/pull/26847), or bring those changes in here
- [x] Check CI across all repos and address any issues

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: package.json
Comment (ceyhun) [inline package.json]: I think these should be in `devDependencies` and could be even better to have them in `react-native-editor/package.json` if possible.

**Comment by ceyhun** [comment]
@cameronvoell Took a quick look and attempted `npm run native ios` to begin. I'm using Xcode 12.4. 

The first error was during `bundle exec pod install` phase where a greater version of CocoaPods was needed. I upgraded it to 1.10 and pushed the changes.

Then I got another error during build phase in Xcode:



Adding `#import <React/RCTEventDispatcher.h>` in this file resolved that issue. 
Warning ⚠️ We still need to commit this change to our `react-native-gesture-handler` fork and update its ref in this PR  

Then I got a `Cycle inside FBReactNativeSpec` error. It seems to be this issue: https://github.com/facebook/react-native/issues/31034. Using [this](https://github.com/facebook/react-native/issues/31034#issuecomment-812564390) proposed solution, I was able to resolve it. I also pushed this change.

After that the iOS demo app built successfully for me 🎉

**Comment by cameronvoell** [comment]
> After that the iOS demo app built successfully for me 🎉

@ceyhun Nice! That's great, I only got to test this for a short bit end of day, and was seeing an error, so guessing it's something with my local env since it worked for you. Let me know if you have any ideas, otherwise I spend some more time looking.

Xcode version is 12.5. My error looked like this:
<details>

`npm run native ios` output:
```
** BUILD FAILED **


The following build commands failed:
	CompileC /Users/cameronvoell/Library/Developer/Xcode/DerivedData/GutenbergDemo-eqaoziwexisoucddwaicvmipdjoc/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/RNGestureHandler.build/Objects-normal/x86_64/RNGestureHandlerManager.o /Users/cameronvoell/A8C/DEVELOP/gutenberg/node_modules/react-native-gesture-handler/ios/RNGestureHandlerManager.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
(6 failures)
```

In Xcode:
image


</details>

**Comment by ceyhun** [comment]
@cameronvoell That seems to be the same error I shared above (only the screenshot is in light mode) 😄 Please check my comment above to resolve it.

**Comment by hypest** [comment]
Headsup that [React Native v0.64.1 was released yesterday that includes a fix for Xcode 12.5](https://github.com/facebook/react-native/releases/tag/v0.64.1).

**Comment by ceyhun** [comment]
Another heads up :) Xcode 12.5 requires macOS 11, which is in [private preview](https://github.com/actions/virtual-environments/issues/2486) on GitHub Actions and not available on CircleCI.

**Comment by cameronvoell** [comment]
Hi @ceyhun , I got the ios demo app running from Xcode (still seeing errors when runnning `npm run native ios` command) but the editor is displaying only the inserter and menu options with the Block-list missing. I'm on Xcode 12.4 and testing on a iPhone 11 simulator. Can you confirm if you are seeing the same? 

| visual editor | html editor |
| -------- | -------- |
|  | |

**Comment by ceyhun** [comment]
`npm run native ios` seems to throw an error for me as well. But it runs `react-native run-ios` in the end and 
the underlying command `xcodebuild -workspace GutenbergDemo.xcworkspace -configuration Debug -scheme GutenbergDemo -destination id=...` 
works fine for me. Same for building it directly from Xcode. But then the white screen you shared is there for me too.

I tried to move `dependencies` in `package.json` to `packages/react-native-editor/package.json` first.
The error I'm currently getting is in the script phase `[CP-User] Generate Specs` of `FBReactNativeSpec` dependency:

```
nvm is not compatible with the "npm_config_prefix" environment variable: currently set to "....gutenberg/packages/react-native-editor"
Run `unset npm_config_prefix` to unset it.
```

I tried adding `. "$HOME/.nvm/nvm.sh"` to the top of the `Bundle React Native code and images` script phase of GutenbergDemo target, but that didn't fix the issue.

I pushed the `package.json` changes to this branch. I'll keep looking into this.

**Comment by ceyhun** [comment]
I fixed the `unset npm_config_prefix` error with another workaround in the Podfile. After that `npm run native ios` started working again :tada:

But in the simulator I started seeing errors saying `css-to-react-native-transform` is missing. Which is a dependency of `react-native-sass-transformer`, which was missing. I tried to re-install `@wordpress/react-native-editor` to bring back the missing dependencies using these commands:
```
npm uninstall @wordpress/react-native-editor --save && npm install ./packages/react-native-editor --save
``` 

But during the `npm install`, I got multiple TypeScript errors related to the build of `react-native-svg`. I realised that because of those errors, `@wordpress/react-native-editor` wasn't properly installed before as well. 

So I went back to our fork of `react-native-svg` to resolve this issue. Reverting the `@types/react-native` back to what it was before, fixed it. Created [a PR](https://github.com/wordpress-mobile/react-native-svg/pull/12) for that change and linked it in the description here.


After these changes, the white screen issue was still there. I saw in metro logs a warning mentioning `KeyboardAwareFlatList`:
```
Warning: Functions are not valid as a React child. This may happen if you return a Component instead of <Component /> from render.
``` 


After checking `react-native-keyboard-aware-scroll-view`, I came across [this fix](https://github.com/APSL/react-native-keyboard-aware-scroll-view/pull/442). I applied the same change in our fork and created [a PR](https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/pull/12) and linked in the description again. Now the editor is back again 🎉🎊🥳🎆    

I didn't have a chance to properly test if everything is working yet, but I noticed there are some dependencies in `/packages/react-native-editor/node_modules` folder that are not hoisted (`metro-react-native-babel-preset` for example), because they are defined in more than one `package.json` with different versions. It would be a good idea to try to unify them under the top-level `node_modules` folder by having them only in a single `package.json`.

**Comment by cameronvoell** [comment]
I noticed today that the Android demo build was broken after we updated the `react-native-gesture-handler`, but I found a fix for that on another [branch](https://github.com/wordpress-mobile/react-native-gesture-handler/tree/upgrade-to-64) from @dratwas (thanks!). On that branch `react-native-gesture-handler` was caught up with upstream fork, and I noticed it built the ios demo without requiring the `#import <React/RCTEventDispatcher.h>` update. 

~~The only issue I'm seeing now is a an issue with focusing on RichText in Android (I know we've had to fix similar bugs in the past, but couldnt find a quick easy fix yet). The error that shows up when clicking on text on Android and iOS is :
I think the only place `focusTextInput` is called is [here](https://github.com/WordPress/gutenberg/blob/fbf0eb266d5a55b420cf7be6e4cc449a64ff8307/packages/react-native-aztec/src/AztecView.js#L187). Maybe you have some ideas on how to fix the focus in text issue on Android?~~ I found where I [accidentally overrode](https://github.com/WordPress/gutenberg/pull/29118/commits/da3608194d074bf918cb11f5cd236c5b5a28139b#diff-166bf22d90bde8c6ae37050a20ead06a22606d7f81ad6e48963269e0736e3f83) a previous fix while resolving conflicts during catching up with trunk. I [fixed that](https://github.com/WordPress/gutenberg/pull/29118/commits/41968e2bacdf7971606dd09d3abfb148d1c15181) and now Android demo app is looking good!

**Comment by hypest** [comment]
I restarted the "Compressed Size / Check (14) (pull_request)" job even though it was successful already, to see if it still succeeds. That job fails in https://github.com/WordPress/gutenberg/pull/32517 which is not clear to me yet why.

File: .github/workflows/create-block.yml
Comment (gziolo) [inline .github/workflows/create-block.yml]: Can you explain why is `npm install` necessary with React 0.64?

Comment (ceyhun) [reply .github/workflows/create-block.yml]: Sorry, I should've mentioned this in the description, will add it there shortly after.

`npm ci` was throwing `cb() never called!` errors randomly both on Linux and macOS on GitHub Actions. We're still not sure what's the reason. 
I tried various different combinations of node, npm, caching, etc. also debugging manually. Those attempts can be traced in some of [these commits](https://github.com/WordPress/gutenberg/commits/rnmobile/try/upgrade-0-64?after=436975285c0002e08cc33b62662e7549eff1e574+34&author=ceyhun&branch=rnmobile%2Ftry%2Fupgrade-0-64). Only thing that made the `cb() never called!` errors go away so far was using `npm install` instead of `npm ci`. I believe this might be an npm bug and hope it will be fixed with npm@7.

File: packages/components/src/mobile/picker/index.android.js
Comment (hypest) [inline packages/components/src/mobile/picker/index.android.js:58]: 👋 @ceyhun , looks like [there's this PR](https://github.com/WordPress/gutenberg/pull/27326) that intentionally replaced the Fragment for a View. Do you think we should keep the View?

File: .github/workflows/create-block.yml
Comment (gziolo) [reply .github/workflows/create-block.yml]: @youknowriad and @desrosj, do you know how to make those GitHub actions work by keeping `npm ci` that should be faster?

File: packages/components/src/scroll-lock/test/index.js
Comment (gziolo) [inline packages/components/src/scroll-lock/test/index.js]: Unit tests have fake timers enabled so it might be a coincidence that `setTimeout` works as expected here. More details on the topic:

https://jestjs.io/docs/timer-mocks#advance-timers-by-time

I haven't tested it but it's probably more future-proof to code it as:

```js
jest.advanceTimersByTime( 1 ); // not sure about the time value
expectLocked( false );
```

File: packages/react-native-bridge/.eslintrc.js
Comment (gziolo) [inline packages/react-native-bridge/.eslintrc.js]: It should be moved to the overrides section for RN files:

https://github.com/WordPress/gutenberg/blob/eaa1be08e278a76774c040414c1691058caaa653/.eslintrc.js#L167-L179

This way we can control it from one place and avoid local ESLint config files.

File: packages/react-native-aztec/.eslintrc.js
Comment (gziolo) [inline packages/react-native-aztec/.eslintrc.js]: Same as in https://github.com/WordPress/gutenberg/pull/29118/files#r651667754.

File: packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java
Comment (hypest) [inline packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java:562]: Looks like not using the `receiveCommand(..., int ..., ...)` variant anymore ([which is deprecated alright](receiveCommand)) would let us also remove [the getCommandsMap override](https://github.com/WordPress/gutenberg/blob/436975285c0002e08cc33b62662e7549eff1e574/packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java#L554). No need to do it in this PR, just mentioning it for our future selves. We should open a ticket about it after merging the RN 0.64 upgrade.

File: packages/components/src/scroll-lock/test/index.js
Comment (hypest) [reply packages/components/src/scroll-lock/test/index.js]: cc @vdwijngaert since I see that change introduced in [this commit](https://github.com/WordPress/gutenberg/commit/79b51c9d52e209fd5e51c8480be8962b5596378f) in the [React v17 upgrade PR](https://github.com/WordPress/gutenberg/pull/26847).

**Review by hypest — APPROVED** [review]
Had a look at the native mobile changes and look good to me! I've only left some non-blocking comments.

Also, I didn't particularly review the iOS side changes but I trust our (successful) testing so far for that. At this point, and since the upgrade effort is already monumental, takes long time and is hard to keep up-to-date, I'm fine with optimistically merging and we can fix any issues afterwards.

That said, since a few folks are doing a pass on the web-side of things (I see @gziolo's for example), let's wait for those review passes to conclude as well before merging.

File: packages/components/src/mobile/picker/index.android.js
Comment (ceyhun) [reply packages/components/src/mobile/picker/index.android.js:58]: In that PR the short syntax of Fragment was used like `<> </>`, which didn't support keys and maybe they were not aware of the [keyed Fragments](https://reactjs.org/docs/fragments.html#keyed-fragments). And also another issue that was solved in that PR, which was using `index` as a key and that was also in the commit in which this change was first introduced [here](https://github.com/WordPress/gutenberg/commit/d40168833531cbcc371c186864ebe3f91f44f14d#diff-9daa2ef947ebc2f02763111c4ffcac3a4f67c13f5c7ead89294ab24c10ed4a99). 
I think adding an extra `View` to the component tree shouldn't be necessary and `Fragment` should be fine.

File: packages/components/src/scroll-lock/test/index.js
Comment (ceyhun) [reply packages/components/src/scroll-lock/test/index.js]: I don't remember this test failing even before I merged in the React 17 upgrade branch, maybe it's because I had installed a more recent version of [@wojtekmaj/enzyme-adapter-react-17](https://github.com/wojtekmaj/enzyme-adapter-react-17) (0.6.2 instead of 0.3.1). I'll try reverting this commit altogether to see if it works fine.

**Comment by gziolo** [comment]
From my perspective, the only concern for merging this PR is `npm ci` replaced in `npm install` in GitHub actions. It's something that we could change to move forward but we need to have a clear idea why it fails and how we can resolve it.

File: packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java
Comment (ceyhun) [reply packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java:562]: Sounds good 👍 I have limited knowledge around here so I only applied the minimum changes necessary to make this work, but I agree that this file might need some further refactoring.

File: packages/components/src/scroll-lock/test/index.js
Comment (ceyhun) [reply packages/components/src/scroll-lock/test/index.js]: Reverting the commit didn't work :) But using `jest.advanceTimersByTime` instead worked fine. [`3e095a7` (#29118)](https://github.com/WordPress/gutenberg/pull/29118/commits/3e095a70589f5a3dc64a309d6159875bc742ef29)

File: .github/workflows/create-block.yml
Comment (youknowriad) [reply .github/workflows/create-block.yml]: I'm not sure this change is what caused the test to start passing, @desrosj made a change on trunk to improve the situation for these failures already. I think we still have some cases left but less than before.

**Comment by youknowriad** [comment]
> From my perspective, the only concern for merging this PR is npm ci replaced in npm install in GitHub actions.

Can we just revert that and see, I think the main problem for these failures is independent from the upgrade.

File: packages/components/src/scroll-lock/test/index.js
Comment (gziolo) [reply packages/components/src/scroll-lock/test/index.js]: I'm happy but surprised it worked on first try 🤣

**Comment by youknowriad** [comment]
Note that the two failures we're seeing now (performance and post title) are unrelated for this PR and exists on trunk and we're working on them in #32686

**Review by gziolo — APPROVED** [review]
Great work, assuming that `npm ci` works the same way as other PRs at the moment, this patch looks good to go.

Amazing teamwork everyone involved to get RN 0.64 and React 17 upgrades working 🎉 👏🏻

**Comment by ceyhun** [comment]
> Note that the two failures we're seeing now (performance and post title) are unrelated for this PR and exists on trunk and we're working on them in #32686

Great to hear that 😅 I've pushed 2 commits (https://github.com/WordPress/gutenberg/pull/29118/commits/85c4331a944a6db374500f9faaa418ca8b741138, https://github.com/WordPress/gutenberg/pull/29118/commits/e8cd4af6dfc63f64e2fe20b862c6a99b35ab1b1a) replacing `npm install` with `npm ci` to how they were before. Didn't see any `cb() never called!` errors, but I fear that might be related to `~/.npm` cache still being in place and maybe the error will return when a new change to `package-lock.json` that makes the cache invalid. In any case I think we can go back to `npm install` or try to find another solution if that ever happens.

I'll get one final merge from `trunk` to fix remaining checks as https://github.com/WordPress/gutenberg/pull/32686 seems to have been merged to `trunk`.

File: packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java
Comment (ceyhun) [reply packages/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java:562]: Created an issue for that: https://github.com/wordpress-mobile/gutenberg-mobile/issues/3635

**Pull request merged**
